ovirt_qos, ovirt_disk_profile, ovirt_disk: Add modules to allow for creation and updating of disk_profiles

odonnelln opened the pull request:
Summary

- Adds ovirt_qos module to create a new QoS entry
- Adds ovirt_disk_profile to create a new disk profile
- Updates ovirt_disk to allow you to define the disk_profile on creation or to update a disk that currently exists

Examples
```yaml
  ovirt.ovirt.ovirt_qos:
    auth: "{{ ovirt_auth }}"
    name: "myqos"
    data_center: "Default"
    state: "present"
    max_iops: "100"
    max_throughput: "100"
    type: "storage"

  ovirt.ovirt.ovirt_disk_profile:
    auth: "{{ ovirt_auth }}"
    name: "mydisk"
    data_center: "Default"
    state: "present"
    storage_domain: "mystoragedomain"
    qos: "myqos"

  ovirt.ovirt.ovirt_disk:
    auth: "{{ ovirt_auth }}"
    name: "mydisk"
    profile: "myqos"
    state: "present"

```

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+DOCUMENTATION = '''
+---
+module: ovirt_disk_profile
+short_description: "Module to manage storage domain disk profiles in oVirt/RHV"
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+author:
+- "Niall O Donnell (niall.odonnell@ppb.com)"
+description:
+    - "Module to manage storage domain disk profiles in oVirt/RHV."
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+        type: str
+    qos:
+        description:
+            - "Name of the QoS entry on the disk profile. If not passed defaults to oVirt/RHV HE default"
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+DOCUMENTATION = '''
+---
+module: ovirt_qos
+short_description: "Module to manage QoS entries in oVirt/RHV"
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+author:
+- "Niall O Donnell (niall.odonnell@ppb.com)"
+description:
+    - "Module to manage QoS entries in oVirt/RHV."
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+    type: str
+    sample: 7de90f31-222c-436c-a1ca-7e655bd5b60c
+qos:
+    description: "Dictionary of all the QoS attributes. QoS attributes can be found on your oVirt/RHV instance
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+    type: str
+    sample: 7de90f31-222c-436c-a1ca-7e655bd5b60c
+disk_profile:
+    description: "Dictionary of all the disk profile attributes. Disk profile attributes can be found on your oVirt/RHV instance
```
Please remove the `/RHV` or replace the `oVirt/RHV` with `@NAME@`.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+        description:
+            - "Name of the data center where the QoS entry should be created."
+        type: str
+        max_iops=dict(default=None),
+        read_iops=dict(default=None),
+        write_iops=dict(default=None),
+        max_throughput=dict(default=None),
+        read_throughput=dict(default=None),
+        write_throughput=dict(default=None),
```
Please remove these lines the documentation is not compiled properly because of it.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+    data_center: "Default"
+    name: "test_qos_01"
+    state: "present"
+    type: "storage"
+
+# Create a new storage QoS with default values for max_iops and read_throughput but 100 for write throughput
+- ovirt_qos:
+    auth: "{{ ovirt_auth }}"
+    data_center: "Default"
+    name: "test_qos_01"
+    state: "present"
+    type: "storage"
+    write_throughput: 100
+
+# Create a new storage QoS with default values for write_iops and max_throughput but 100 for read iops
+- ovirt_qos:
+    auth: "{{ ovirt_auth }}"
+    data_center: "Default"
+    name: "test_qos_01"
+    state: "present"
+    type: "storage"
+    read_iops: 100
+
+# Create a new storage QoS with 100 max_iops and 200 max_throughput
+- ovirt_qos:
+    auth: "{{ ovirt_auth }}"
+    data_center: "Default"
+    name: "test_qos_01"
+    state: "present"
+    type: "storage"
+    max_iops: 100
+    max_throughput: 100
+
+# Remove a storage QoS
+- ovirt_qos:
+    auth: "{{ ovirt_auth }}"
+    data_center: "Default"
+    name: "test_qos_01"
+    state: "absent"
+    type: "storage"
```
Please add the prefix `@NAMESPACE@.@NAME@` to the `ovirt_qos` name.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+            return otypes.QosType.HOSTNETWORK
+        elif type == 'cpu':
+            return otypes.QosType.CPU
+        else:
+            return None
```
Please remove the `else` and just leave `return None`.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+    if dc is None:
+        dc = get_entity(dcs_service.service(dc_name))
+        if dc is None:
+            return None
```
Please raise the exception that the Datacenter was not found.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+    get_entity
+)
+
+class DiskProfileModule(BaseModule):
```
Please add an extra line.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+        return storage_domain
+
+
+    def build_entity(self):
```
Please remove the extra line.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+
+
+
+def main():
```
Please remove the extra 2 lines.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+    finally:
+        connection.close(logout=auth.get('token') is None)
+
+if __name__ == "__main__":
```
Please add an extra line.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+
+if __name__ == "__main__":
+  main()
+
```
remove the blank line at end of file

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+
+if __name__ == "__main__":
+  main()
+
```
remove the blank line at end of file

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+    finally:
+        connection.close(logout=auth.get('token') is None)
+
+if __name__ == "__main__":
```
Please add an extra line.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+        argument_spec=argument_spec,
+        required_one_of=[['id', 'name']],
+        mutually_exclusive=[
+            ['max_iops','read_iops'],
```
Add whitespace after ,

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+            ['max_iops','read_iops'],
+            ['max_iops', 'write_iops'],
+            ['max_throughput','read_throughput'],
+            ['max_throughput', 'write_througput']
```
Add whitespace after `,`

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+    get_entity
+)
+
+class QosModule(BaseModule):
```
Please add an extra line.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+            max_write_throughput=int(self._module.params.get('write_throughput')) if self._module.params.get('write_throughput') is not None else None,
+        )
+
+def _get_qoss_service(connection, dc_name):
```
Please add an extra line.

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+
+
+
+def main():
```
Please remove the extra line.

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/python
```
Add please to the shebang  license.
```
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2022 Red Hat, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/python
```
Please add license
```
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2022 Red Hat, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+        max_throughput=dict(default=None),
+        read_throughput=dict(default=None),
+        write_throughput=dict(default=None),
+    max_iops:
+        description:
+            - "The max number of read/write iops. If passed you can't pass a value for C(read_iops) or C(write_iops)"
+            - "If no value is given it will default to the HE value, assuming C(read_iops) or C(write_iops) hasn't been set"
+        type: str
+    write_iops:
+        description:
+            - "The max number of write iops. If passed you can't pass a value for C(max_iops)"
+            - "If no value is given it will default to the HE value, assuming C(max_iops) hasn't been set"
+        type: str
+    read_iops:
+        description:
+            - "The max number of read iops. If passed you can't pass a value for C(max_iops)"
+            - "If no value is given it will default to the HE value, assuming C(max_iops) hasn't been set"
+        type: str
+    max_throughput:
+        description:
+            - "The max number of read/write throughput. If passed you can't pass a value for C(read_throughput) or C(write_throughput)"
+            - "If no value is given it will default to the HE value, assuming C(read_throughput) or C(write_throughput) hasn't been set"
+        type: str
+    write_throughput:
+        description:
+            - "The max number of write throughput. If passed you can't pass a value for C(max_throughput)"
+            - "If no value is given it will default to the HE value, assuming C(max_throughput) hasn't been set"
+        type: str
+    read_throughput:
+        description:
+            - "The max number of read throughput. If passed you can't pass a value for C(max_throughput)"
+            - "If no value is given it will default to the HE value, assuming C(max_throughput) hasn't been set"
+        type: str
```
Shouldn't these be integers?

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+
+    :returns: ovirt.services.QossService or None
+    """
+    dcs_service = connection.system_service().data_centers_service()
+    dc = search_by_name(dcs_service, dc_name)
+
+    if dc is None:
+        dc = get_entity(dcs_service.service(dc_name))
+        if dc is None:
+            return None
+
+    dc_service = dcs_service.data_center_service(dc.id)
```
Maybe could be replace with
```python
dcs_service = connection.system_service().data_centers_service()
return dcs_service.data_center_service(get_id_by_name(dcs_service, dc_name)).qoss_service()
```

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+
+        :return: otypes.StorageDomain or None
+        """
+        storage_domain_name = self._module.params.get('storage_domain')
+        storage_domains_service = self._connection.system_service().storage_domains_service()
+        storage_domain = search_by_name(storage_domains_service, storage_domain_name)
+
+        if storage_domain is None:
+            storage_domain = get_entity(storage_domains_service.service(storage_domain_name))
+            if storage_domain is None:
+                return None
+
+        return storage_domain
```
Maybe could be replaced by `get_id_by_name` as in comment on QoS

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+
+        :return: otypes.QoS or None
+        """
+        dc_name = self._module.params.get('data_center')
+        dcs_service = self._connection.system_service().data_centers_service()
+        dc = search_by_name(dcs_service, dc_name)
+
+        if dc is None:
+            dc = get_entity(dcs_service.service(dc_name))
+            if dc is None:
+                return None
+
+        dc_service = dcs_service.data_center_service(dc.id)
+        qos_service = dc_service.qoss_service()
+        qos = search_by_name(qos_service, self._module.params.get('qos'))
+
+        if qos is None:
+            qos = get_entity(qos_service.service(self._module.params.get('qos')))
+            if qos is None:
+                return None
+
+        return qos
```
Maybe could be replaced by `get_id_by_name` as in comment on QoS

mnecas commented on plugins/modules/ovirt_disk_profile.py:
```diff
@@ -0,0 +1,225 @@
+                f"The storage domain: {self._module.params.get('storage_domain')} does not exist."
+            )
+        return otypes.DiskProfile(
+            name=self._module.params.get('id') if self._module.params.get('id') else self._module.params.get('name'),
```
The DIskProfile has the id and name. Please change the code to 
```python
name=self._module.params.get('id') if self._module.params.get('name') else None,
id = self._module.params.get('id') if self._module.params.get('id') else None
 ```

mnecas commented on plugins/modules/ovirt_qos.py:
```diff
@@ -0,0 +1,264 @@
+        :return: otypes.QoS
+        """
+        return otypes.Qos(
+            name=self._module.params.get('id') if self._module.params.get('id') else self._module.params.get('name'),
```
The Qos has the id and name. Please change the code to 
```python
name=self._module.params.get('id') if self._module.params.get('name') else None,
id = self._module.params.get('id') if self._module.params.get('id') else None
 ```

mnecas commented on plugins/modules/ovirt_disk.py:
```diff
@@ -716,6 +719,36 @@ def update_storage_domains(self, disk_id):
 
         return changed
 
+    def update_disk_profile(self, disk_id):
+        """
+        Updates the disk_profile for a given disk
+
+        :param disk_id: id of the disk
+        :return: Wether a change has been made to the disk or not
+        """
+        changed = False
+        disk_service = self._service.service(disk_id)
+        disk = disk_service.get()
+        dp_service = self._connection.system_service().disk_profiles_service()
+
+        if self._module.params['profile']:
+            new_profile_id = get_id_by_name(dp_service, self._module.params['profile'])
+            if new_profile_id == disk.disk_profile.id:
+                return changed
+            update_disk = otypes.Disk(
+                id=disk.id,
+                disk_profile=otypes.DiskProfile(id=new_profile_id),
+            )
+            changed = self.action(
+                action='update',
+                disk=update_disk,
+                action_condition=lambda d: new_profile_id != d.disk_profile.id,
+                wait_condition=lambda d: d.status == otypes.DiskStatus.OK,
+                post_action=lambda _: time.sleep(self._module.params['poll_interval']),
+            )['changed']
+
+        return changed
```
Is it necessary to have a separate method for an update of the disk profiles?
I think it should go to the `update_check`
https://github.com/oVirt/ovirt-ansible-collection/blob/master/plugins/modules/ovirt_disk.py#L719

mnecas reviewed (changes requested):
Hi @odonnelln, thanks for the PR great work! I have just some nits, quite a lot of them are repetitive because of the amount of added code, and a lot of them are from the CI. Please don't hesitate to ping me.

mwperina commented:
Hi Niall,

Thanks a lot for contribution. Your patch looks good, but need a bit more work. Could you please take a look at comments?

Thanks,
Martin

odonnelln commented:
@mnecas Sorry for the delay in getting back to this. I have addressed all your comments. Not sure what process is here for resolving comments, in my work we leave it to the person who adds the comment to resolve but please let me know if you want me to resolve them all.

mnecas commented:
@odonnelln once again great job and thank you for you PR!
I added QoS params for network, hostnetwork, cpu, and some details.
